### DISCUSSION TITLE
Issue with horiz_time_volume update_data on sparser intervals (e.g., 30m)

### DISCUSSION THREAD

**Discussion by Kassad1979** [highfestiva/finplot#540] [Q&A]
Hello,

I’m encountering an issue when using `update_data()` with `horiz_time_volume` in finplot. The problem appears specifically when dealing with sparser intervals and certain grouping parameters. Here’s a detailed summary:

1. **Context**:  
   - I’m using `horiz_time_volume()` to plot a volume profile. The initial data and subsequent updates follow the required input format:
     ```python
     [
       (timestamp0, [(price0, volume0), (price1, volume1), ...]),
       (timestamp1, [(price0, volume0), (price1, volume1), ...]),
       ...
     ]
     ```
   - For denser intervals like '1m' or '5m', even with a large amount of data, `update_data()` works flawlessly. No indexing or out-of-bounds errors occur.

2. **The Problem**:  
   - When using sparser intervals like '30m' (or lower frequency) combined with daily grouping (`period='D'`) and similar parameters for bins (e.g., 150 bins), `update_data()` frequently fails.
   - The error typically reports something like:
     ```
     index 1330304400000000000 is out of bounds for axis 0 with size ...
     ```
     indicating that finplot is seemingly interpreting a timestamp as an integer index.

3. **Observations**:  
   - With '1m' or '5m' intervals and `period='D'`, everything is fine. Larger datasets and more dense data result in no errors.
   - With '30m', where each daily aggregation yields fewer data points per day, the issue arises. It seems that with fewer points per grouping period, finplot may struggle internally to map timestamps and volume bins correctly.
   - Adjusting the number of bins or changing intervals hasn’t resolved the issue.  
   - Using the same code and logic, just switching between intervals ('5m' works, '30m' fails) reveals that the underlying logic might depend on having a certain density of data per time period.

4. **What I’ve Tried**:  
   - Changing bins and period parameters.
   - Providing more or less data.
   - Testing incrementally smaller updates.
   - For more granular intervals, `update_data()` works perfectly, which suggests that the format and logic of the code are correct.

5. **Hypothesis**:  
   - `horiz_time_volume` might require a minimum density of data points within each grouped period (day, in this case) to correctly map timestamps to its internal data structure.
   - When the interval is sparser (e.g., '30m') and there are fewer points per day, finplot’s internal indexing or bin mapping could fail, causing the out-of-bounds error.

6. **Request**:  
   - Could you clarify if `horiz_time_volume` and `update_data()` have certain assumptions about data density or continuity?
   - Is there a known limitation or expected workaround for sparser intervals with daily grouping?
   - Any guidance on how to handle fewer data points per period (e.g., ensuring continuous data or using a different period grouping) would be greatly appreciated.

Thank you in advance for your assistance, and for providing such a useful plotting library.

Best regards,  
Marcin

**Comment by highfestiva** [comment]
This is a bug in finplot, sorry about that. I need a fully working, minimal example so I can reproduce the bug. I promise to fix it if you get me that. As you probably can tell, that part of finplot is a hack, and not super-easy to work with... :-/

Thank you for your thanks! :)
